### ISSUE TITLE
U_MISSING_RESOURCE_ERROR with ICU 60.2 c.f. #3193

### ISSUE THREAD

**Issue by oglord** [mapnik/mapnik#3994]
Having read issue #3193 I have tried various things to prevent the same error on Fedora 28. I have compiled mapnik and node-mapnik from source (both master and earlier releases); I have installed node-mapnik from npm; I have installed both packages from the FC28 RPMs, but still get the same ICU-releated errors as described in #3193. It seems that the solution to that problem was making the ICU .dat files available to mapnik, but ICU 60.2 does not have any.

```
module.exports.env = {
    'ICU_DATA':      '/usr/share/icu/60.2',
    'GDAL_DATA':     '/usr/share/gdal',
    'PROJ_LIB':      '/usr/share/proj'
};
```

Yet, /usr/share/icu/60.2 is empty. I have tried placing /usr/lib64/firefox/icudt60l.dat from the same FC28 installation into the ICU_DATA directory referenced above, but to no avail.

**Comment by talaj** [comment]
What `icu-config` is showing?

```console
$ icu-config  --icudata --icudatadir
icudt62l /usr/share/icu/62.1 
```

**Comment by oglord** [comment]
> icu-config --icudata --icudatadir

`icudt60l /usr/share/icu/60.2 
`

**Comment by oglord** [comment]
Is there a reference platform for mapnik and node-mapnik? I have succesfully used the Python bindings on CentOS 7 and Fedora 28, but have so far failed to find a platform on which the nodejs bindings work correctly.

**Comment by talaj** [comment]
I think there is no such thing as reference platform for Mapnik. Recent versions of Mapnik are packaged in Debian or Ubuntu. Openstreetmaps.org is using Mapnik from Ubuntu Bionic, for example.

Did you try to install node-mapnik by npm? It should handle all dependencies. No need to complile from source.

**Comment by tomhughes** [comment]
I'm the Fedora maintainer for mapnik and node-mapnik - do you have a simple reproducer for the nodejs bindings?

**Comment by tomhughes** [comment]
Note that I don't think `ICU_DATA` actually matters on Fedora because I don't think ICU is built to load data via a `dat` file - that's why `/usr/share/icu/60.2` is largely empty. Rather I believe the data is compiled into the shared library.

**Comment by oglord** [comment]
> I'm the Fedora maintainer for mapnik and node-mapnik - do you have a simple reproducer for the nodejs bindings?

The most recent thing I tried was to compile both from the master branch. This is what was linked into mapnik.node:
``
	linux-vdso.so.1 (0x00007ffecb3d7000)
	libmapnik.so.3.1 => /lib/libmapnik.so.3.1 (0x00007fc9c9aee000)
	libboost_filesystem.so.1.66.0 => /lib64/libboost_filesystem.so.1.66.0 (0x00007fc9c98d4000)
	libboost_regex.so.1.66.0 => /lib64/libboost_regex.so.1.66.0 (0x00007fc9c95b7000)
	libcairo.so.2 => /lib64/libcairo.so.2 (0x00007fc9c9297000)
	libpng16.so.16 => /lib64/libpng16.so.16 (0x00007fc9c9063000)
	libproj.so.12 => /lib64/libproj.so.12 (0x00007fc9c8df9000)
	libtiff.so.5 => /lib64/libtiff.so.5 (0x00007fc9c8b81000)
	libwebp.so.7 => /lib64/libwebp.so.7 (0x00007fc9c8914000)
	libxml2.so.2 => /lib64/libxml2.so.2 (0x00007fc9c85ad000)
	libicui18n.so.60 => /lib64/libicui18n.so.60 (0x00007fc9c80ee000)
	libboost_system.so.1.66.0 => /lib64/libboost_system.so.1.66.0 (0x00007fc9c7ee9000)
	libharfbuzz.so.0 => /lib64/libharfbuzz.so.0 (0x00007fc9c7c44000)
	libjpeg.so.62 => /lib64/libjpeg.so.62 (0x00007fc9c79dc000)
	libicuuc.so.60 => /lib64/libicuuc.so.60 (0x00007fc9c761a000)
	libfreetype.so.6 => /lib64/libfreetype.so.6 (0x00007fc9c7363000)
	libz.so.1 => /lib64/libz.so.1 (0x00007fc9c714c000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007fc9c6f48000)
	libstdc++.so.6 => /lib64/libstdc++.so.6 (0x00007fc9c6bb6000)
	libm.so.6 => /lib64/libm.so.6 (0x00007fc9c6822000)
	libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007fc9c660a000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007fc9c63eb000)
	libc.so.6 => /lib64/libc.so.6 (0x00007fc9c602c000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fc9cad96000)
	librt.so.1 => /lib64/librt.so.1 (0x00007fc9c5e24000)
	libicudata.so.60 => /lib64/libicudata.so.60 (0x00007fc9c427b000)
	libpixman-1.so.0 => /lib64/libpixman-1.so.0 (0x00007fc9c3fd6000)
	libfontconfig.so.1 => /lib64/libfontconfig.so.1 (0x00007fc9c3d93000)
	libxcb-shm.so.0 => /lib64/libxcb-shm.so.0 (0x00007fc9c3b8f000)
	libxcb.so.1 => /lib64/libxcb.so.1 (0x00007fc9c3966000)
	libxcb-render.so.0 => /lib64/libxcb-render.so.0 (0x00007fc9c3758000)
	libXrender.so.1 => /lib64/libXrender.so.1 (0x00007fc9c354d000)
	libX11.so.6 => /lib64/libX11.so.6 (0x00007fc9c320c000)
	libXext.so.6 => /lib64/libXext.so.6 (0x00007fc9c2ffa000)
	libjbig.so.2.1 => /lib64/libjbig.so.2.1 (0x00007fc9c2dee000)
	liblzma.so.5 => /lib64/liblzma.so.5 (0x00007fc9c2bc7000)
	libglib-2.0.so.0 => /lib64/libglib-2.0.so.0 (0x00007fc9c28b0000)
	libgraphite2.so.3 => /lib64/libgraphite2.so.3 (0x00007fc9c2684000)
	libbz2.so.1 => /lib64/libbz2.so.1 (0x00007fc9c2473000)
	libexpat.so.1 => /lib64/libexpat.so.1 (0x00007fc9c2238000)
	libuuid.so.1 => /lib64/libuuid.so.1 (0x00007fc9c2031000)
	libXau.so.6 => /lib64/libXau.so.6 (0x00007fc9c1e2d000)
	libpcre.so.1 => /lib64/libpcre.so.1 (0x00007fc9c1bbc000)
``

And this is libmapnik itself:

``
	linux-vdso.so.1 (0x00007ffd19369000)
	libboost_filesystem.so.1.66.0 => /lib64/libboost_filesystem.so.1.66.0 (0x00007f903f4a7000)
	libboost_regex.so.1.66.0 => /lib64/libboost_regex.so.1.66.0 (0x00007f903f18a000)
	libcairo.so.2 => /lib64/libcairo.so.2 (0x00007f903ee6a000)
	libpng16.so.16 => /lib64/libpng16.so.16 (0x00007f903ec36000)
	libproj.so.12 => /lib64/libproj.so.12 (0x00007f903e9cc000)
	libtiff.so.5 => /lib64/libtiff.so.5 (0x00007f903e754000)
	libwebp.so.7 => /lib64/libwebp.so.7 (0x00007f903e4e7000)
	libxml2.so.2 => /lib64/libxml2.so.2 (0x00007f903e180000)
	libicui18n.so.60 => /lib64/libicui18n.so.60 (0x00007f903dcc1000)
	libboost_system.so.1.66.0 => /lib64/libboost_system.so.1.66.0 (0x00007f903dabc000)
	libharfbuzz.so.0 => /lib64/libharfbuzz.so.0 (0x00007f903d817000)
	libjpeg.so.62 => /lib64/libjpeg.so.62 (0x00007f903d5af000)
	libicuuc.so.60 => /lib64/libicuuc.so.60 (0x00007f903d1ed000)
	libfreetype.so.6 => /lib64/libfreetype.so.6 (0x00007f903cf36000)
	libz.so.1 => /lib64/libz.so.1 (0x00007f903cd1f000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007f903cb1b000)
	libstdc++.so.6 => /lib64/libstdc++.so.6 (0x00007f903c789000)
	libm.so.6 => /lib64/libm.so.6 (0x00007f903c3f5000)
	libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007f903c1dd000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f903bfbe000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f903bbff000)
	librt.so.1 => /lib64/librt.so.1 (0x00007f903b9f7000)
	libicudata.so.60 => /lib64/libicudata.so.60 (0x00007f9039e4e000)
	libpixman-1.so.0 => /lib64/libpixman-1.so.0 (0x00007f9039ba9000)
	libfontconfig.so.1 => /lib64/libfontconfig.so.1 (0x00007f9039966000)
	libxcb-shm.so.0 => /lib64/libxcb-shm.so.0 (0x00007f9039762000)
	libxcb.so.1 => /lib64/libxcb.so.1 (0x00007f9039539000)
	libxcb-render.so.0 => /lib64/libxcb-render.so.0 (0x00007f903932b000)
	libXrender.so.1 => /lib64/libXrender.so.1 (0x00007f9039120000)
	libX11.so.6 => /lib64/libX11.so.6 (0x00007f9038ddf000)
	libXext.so.6 => /lib64/libXext.so.6 (0x00007f9038bcd000)
	libjbig.so.2.1 => /lib64/libjbig.so.2.1 (0x00007f90389c1000)
	liblzma.so.5 => /lib64/liblzma.so.5 (0x00007f903879a000)
	libglib-2.0.so.0 => /lib64/libglib-2.0.so.0 (0x00007f9038483000)
	libgraphite2.so.3 => /lib64/libgraphite2.so.3 (0x00007f9038257000)
	libbz2.so.1 => /lib64/libbz2.so.1 (0x00007f9038046000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f9040549000)
	libexpat.so.1 => /lib64/libexpat.so.1 (0x00007f9037e0b000)
	libuuid.so.1 => /lib64/libuuid.so.1 (0x00007f9037c04000)
	libXau.so.6 => /lib64/libXau.so.6 (0x00007f9037a00000)
	libpcre.so.1 => /lib64/libpcre.so.1 (0x00007f903778f000)
``

**Comment by oglord** [comment]
> Note that I don't think `ICU_DATA` actually matters on Fedora because I don't think ICU is built to load data via a `dat` file - that's why `/usr/share/icu/60.2` is largely empty. Rather I believe the data is compiled into the shared library.

Indeed, which is why the fix for locating the correct `dat` file did not work for me.

**Comment by tomhughes** [comment]
Yes I can run ldd on the library... What I mean is do you have a bit of javascript I can run (ideally with needing a database loaded...) that will demonstrate the problem?

**Comment by oglord** [comment]
> I think there is no such thing as reference platform for Mapnik. Recent versions of Mapnik are packaged in Debian or Ubuntu. Openstreetmaps.org is using Mapnik from Ubuntu Bionic, for example.
> 
> Did you try to install node-mapnik by npm? It should handle all dependencies. No need to complile from source.

Yes - the dependencies were all found, but running the generator script resulted in the same `could not create BreakIterator: U_MISSING_RESOURCE_ERROR` runtime error as when I compiled from source.

**Comment by oglord** [comment]
> Yes I can run ldd on the library... What I mean is do you have a bit of javascript I can run (ideally with needing a database loaded...) that will demonstrate the problem?

[generate.txt](https://github.com/mapnik/mapnik/files/2460141/generate.txt)

**Comment by tomhughes** [comment]
What's an "appropriate" style sheet though? This is very likely to be sensitive to the style and the exact data being used... I don't have a ready to go rendering environment which is why I was hoping you had a minimal test case but apparently you don't.

**Comment by oglord** [comment]
> What's an "appropriate" style sheet though? This is very likely to be sensitive to the style and the exact data being used... I don't have a ready to go rendering environment which is why I was hoping you had a minimal test case but apparently you don't.

The stylesheet I use is based on the example provided on the OSM Wiki for Mapnik 2.2. I will create a ZIPped version of it to attach, because it includes all the `.xml.inc` bits that define XML entities and suchlike.

**Comment by oglord** [comment]
[xml.zip](https://github.com/mapnik/mapnik/files/2463849/xml.zip) is a ZIP file of the XML used to reproduce the error. The Shapefiles layer has been commented-out to prevent needing to upload megabytes of SHPs, but the error is generated regardless.
